[dpkg_divert] Convert all roles to use the module

Alphix opened the pull request:
I've tested all roles that are affected by this PR (great fun) except for `postgresql_server` (the autobackup tool doesn't exist in Bullseye)

Alphix committed:
[dpkg_divert] Update documentation

The description of the default suffix has not been correct since
commit 60af5a8baf26969f437b81ad23de45514654cb3f, which added
different contextual suffixes.

Alphix committed:
[backup2l] Use the dpkg_divert module

Alphix committed:
[logrotate] Use the dpkg_divert module

Alphix committed:
[ntp] Use the dpkg_divert module

Alphix committed:
[libvirtd] Use the dpkg_divert module

Alphix committed:
[imapproxy] Use the dpkg_divert module

Alphix committed:
[ldap] Use the dpkg_divert module

Alphix committed:
[libuser] Use the dpkg_divert module

Alphix committed:
[avahi] Use the dpkg_divert module

Alphix committed:
[apache] Use the dpkg_divert module

Alphix committed:
[apt] Use the dpkg_divert module

Alphix committed:
[apt_cacher_ng] Use the dpkg_divert module

Alphix committed:
[debops_legacy] Use the dpkg_divert module

Alphix committed:
[dnsmasq] Use the dpkg_divert module

Alphix committed:
[docker_registry] Use the dpkg_divert module

Alphix committed:
[elasticsearch] Use the dpkg_divert module

Alphix committed:
[etc_services] Use the dpkg_divert module

Alphix committed:
[etckeeper] Use the dpkg_divert module

Alphix committed:
[fail2ban] Use the dpkg_divert module

Alphix committed:
[ferm] Use the dpkg_divert module

Alphix committed:
[freeradius] Use the dpkg_divert module

Note that the docs appear to be misleading. The role won't (unless I'm
completely mistaken) check if any files are Debian conffiles, it just
checks if there are existing diversions.

Alphix committed:
[fuse] Use the dpkg_divert module

Alphix committed:
[grub] Use the dpkg_divert module

Alphix committed:
[icinga] Use the dpkg_divert module

Alphix committed:
[ifupdown] Use the dpkg_divert module

Alphix committed:
[kibana] Use the dpkg_divert module

Alphix committed:
[libvirt_qemu] Use the dpkg_divert module

Alphix committed:
[lvm] Use the dpkg_divert module

Alphix committed:
[machine] Use the dpkg_divert module

Alphix committed:
[minidlna] Use the dpkg_divert module

Alphix committed:
[nginx] Use the dpkg_divert module

Alphix committed:
[nscd] Use the dpkg_divert module

Alphix committed:
[opendkim] Use the dpkg_divert module

Alphix committed:
[pam_access] Use the dpkg_divert module

Alphix committed:
[php] Use the dpkg_divert module

Alphix committed:
[postgresql_server] Use the dpkg_divert module

Alphix committed:
[resolvconf] Use the dpkg_divert module

Alphix committed:
[slapd] Use the dpkg_divert module

Alphix committed:
[snapshot_snapper] Use the dpkg_divert module

Alphix committed:
[snmpd] Use the dpkg_divert module

Alphix committed:
[sshd] Use the dpkg_divert module

Alphix committed:
[sysctl] Use the dpkg_divert module

Alphix committed:
[tcpwrappers] Use the dpkg_divert module

Alphix committed:
[unattended_upgrades] Use the dpkg_divert module

Alphix committed:
Make yamllint happy

Alphix committed:
Make Ansible syntax check happy

Alphix committed:
Make ansible-lint happy(ier)

Alphix commented:
Not 100% sure what ansible-lint wants from me :)

drybjed commented on ansible/roles/grub/tasks/main.yml:
```diff
@@ -119,16 +119,17 @@
+    dest: '/etc/grub.d/10_linux'
+    remote_src: True
+    force: False
+  when: (grub__combined_users|length > 0)
```
As an example here, `ansible-lint` wants all `copy` and `template` uses to include the `mode` parameter to explicitly define the file permissions.

drybjed commented on ansible/roles/snapshot_snapper/tasks/main.yml:
```diff
@@ -36,17 +40,18 @@
 
 - name: Copy diverted configuration file to original location
-  command: cp {{ item }}.dpkg-divert {{ item }}
-  args:  # noqa no-handler
```
Since this task shouldn't be done as a handler (well, we could add it as a handler and do meta: flush_handlers but that's not really needed here...), make sure to copy the `# noqa no-handler` comment to tell `ansible-lint` to ignore this issue.

Alphix commented:
Ah, I never got that far because `ansible-lint` errored out, complaining that `dpkg_divert` was an unknown module.

I've [figured out](https://github.com/willthames/ansible-review/issues/16) that setting the `ANSIBLE_LIBRARY` environment variable works around it, so I have the real ansible-lint errors now. I'll go through them and update the PR.

Alphix committed:
More ansible-lint fixes

drybjed commented on ansible/roles/apt/tasks/main.yml:
```diff
@@ -108,12 +100,12 @@
+  dpkg_divert:
+    path: '/etc/apt/sources.list'
+    state: 'absent'
+    delete: True
```
The combination of creation and removal of diversion can be done in a single `dpkg_divert` task - just use the `state` parameter conditionally (the `delete` parameter does nothing with `state: present`). Do you think that you could merge this and other similar tasks into a single task each time this happens? We can leave it for later if it becomes too complex.

Alphix replied:
I'll have a look

Alphix replied:
Ok, done

Alphix committed:
[apt] Merge diversion/reversion

Alphix committed:
[etc_services] Merge diversion/reversion

Alphix committed:
[ferm] Merge diversion/reversion

Alphix committed:
[grub] Merge diversion/reversion

Alphix committed:
[logrotate] Merge diversion/reversion

Alphix committed:
[logrotate] Add logrotate__combined_config to simplify tasks

Alphix committed:
[machine] Merge diversion/reversion

Alphix committed:
[pam_access] Merge diversion/reversion

Alphix committed:
[resolvconf] Merge diversion/reversion

Alphix committed:
[sshd] Merge diversion/reversion

Alphix committed:
[sysctl] Merge diversion/reversion

Alphix committed:
[tcpwrappers] Merge diversion/reversion

Alphix committed:
[unattended_upgrades] Merge diversion/reversion

Alphix committed:
[icinga] Merge diversion/reversion

Alphix committed:
[freeradius] Merge diversion/reversion

Alphix committed:
[freeradius] Add diversion check to file removal task

The original version had a check on "divert_state", which was
a variable that was only defined for a task higher up and which
appears to unintentionally have been a noop.

Alphix commented:
@drybjed : I've merged the divert/revert tasks for all roles except for `apt_cacher_ng` (which has the divert and revert in two different task files....I have a separate PR for that role which I'll submit later, once I've tested it).

I've tested all the roles affected by the merger of divert/revert (rather light testing on `icinga` and `freeradius` given that they don't actually do much diversion/reversion out of the box) and it all seems to work.

Wouldn't suggest backporting this PR given how many roles it touches....

drybjed reviewed (approved):
Excellent work, thanks! This won't be backported due to large number of changes.

ypid commented:
Awesome work! I will review it now. I consider this an internal refactoring. I am not sure if it needs to be in the changelog? The changelog is long enough already ;-)
